File: festival-security/src/main/java/cn/bdqfork/security/util/SecurityUtils.java
package cn.bdqfork.security.util;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.functions.Function;
import io.vertx.reactivex.ext.auth.User;

import java.util.List;

/**
 * @author bdq
 * @since 2020/1/28
 */
public class SecurityUtils {
    public static Observable<Boolean> isPermited(User user, String[] permitsOrRoles) {
        return Observable.fromArray(permitsOrRoles)
                .flatMap(new Function<String, ObservableSource<Boolean>>() {
                    @Override
                    public ObservableSource<Boolean> apply(String role) throws Exception {
                        return user.rxIsAuthorized(role)
                                .toObservable();
                    }
                })
                .toList()
                .map(new Function<List<Boolean>, Boolean>() {
                    @Override
                    public Boolean apply(List<Boolean> results) throws Exception {
                        for (Boolean result : results) {
                            if (!result) {
                                return false;
                            }
                        }
                        return true;
                    }
                }).toObservable();
    }
}

File: festival-kotlin-web/src/main/java/cn/bdqfork/web/util/SecurityUtils.kt
package cn.bdqfork.web.util

import cn.bdqfork.web.constant.LogicType
import io.vertx.ext.auth.User
import io.vertx.kotlin.coroutines.awaitResult
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.runBlocking

/**
 * @author bdq
 * @since 2020/1/28
 */
object SecurityUtils {
    @JvmStatic
    fun isPermited(user: User, permits: Array<out String>, logicType: LogicType): Boolean = runBlocking {
        return@runBlocking GlobalScope.async {
            var finalResult = awaitResult<Boolean> { h -> user.isAuthorized(permits[0], h) }
            for (i in 1 until permits.size) {
                val result = awaitResult<Boolean> { h -> user.isAuthorized(permits[i], h) }
                if (logicType == LogicType.AND) {
                    finalResult = finalResult && result
                }
                if (logicType == LogicType.OR) {
                    finalResult = finalResult || result
                }
            }
            return@async finalResult
        }.await()
    }
}